src/mpi/init/mpir_init.c

```diff
@@ -352,3 +360,28 @@ int MPIR_Finalize_impl(void)
 {
     return MPII_Finalize(NULL);
 }
+
+#ifdef ENABLE_QMPI
+static void qmpi_teardown(void)
+{
+    int counter = 0;
+
+    MPL_free(MPIR_QMPI_pointers);
+    MPL_free(MPIR_QMPI_storage);
+    if (MPIR_CVAR_QMPI_TOOL_LIST != NULL) {
+        /* We allocate one extra tool name so we don't have to shift to avoid the MPICH functions */
+        MPL_free(MPIR_QMPI_tool_names[counter++]);
+        MPL_free(MPIR_QMPI_tool_names[counter++]);
+        size_t len = strlen(MPIR_CVAR_QMPI_TOOL_LIST);
+        for (int i = 0; i <= len; i++) {
+            if (MPIR_CVAR_QMPI_TOOL_LIST[i] == ':') {
+                MPL_free(MPIR_QMPI_tool_names[counter++]);
+            }
+        }
+    }
+    MPL_free(MPIR_QMPI_first_fn_ptrs);
+    MPL_free(MPIR_QMPI_first_tool_ids);
+    MPL_free(MPIR_QMPI_tool_names);
+    MPL_free(MPIR_QMPI_tool_init_callbacks);
+}
```

Comment:
Why `qmpi_teardown` is not defined in `src/mpi_t/qmpi_register.c`? I understand you want to make it a static function, but I think it is more important to group code to their closer context.

Reply:
There was probably a good reason at some point. 🤷 I think you're probably right that moving it makes more sense now.